File: ios/HMSManager.swift
import HMSSDK

@objc(HMSManager)
class HMSManager: RCTEventEmitter {

    var hmsCollection = [String: HMSRNSDK]()

    let ON_PREVIEW = "ON_PREVIEW"
    let ON_JOIN = "ON_JOIN"
    let ON_ROOM_UPDATE = "ON_ROOM_UPDATE"
    let ON_PEER_UPDATE = "3"
    let ON_TRACK_UPDATE = "ON_TRACK_UPDATE"
    let ON_ROLE_CHANGE_REQUEST = "ON_ROLE_CHANGE_REQUEST"
    let ON_CHANGE_TRACK_STATE_REQUEST = "ON_CHANGE_TRACK_STATE_REQUEST"
    let ON_REMOVED_FROM_ROOM = "ON_REMOVED_FROM_ROOM"
    let ON_ERROR = "ON_ERROR"
    let ON_MESSAGE = "ON_MESSAGE"
    let ON_SPEAKER = "ON_SPEAKER"
    let RECONNECTING = "RECONNECTING"
    let RECONNECTED = "RECONNECTED"
    let ON_RTC_STATS = "ON_RTC_STATS"
    let ON_LOCAL_AUDIO_STATS = "ON_LOCAL_AUDIO_STATS"
    let ON_LOCAL_VIDEO_STATS = "ON_LOCAL_VIDEO_STATS"
    let ON_REMOTE_AUDIO_STATS = "ON_REMOTE_AUDIO_STATS"
    let ON_REMOTE_VIDEO_STATS = "ON_REMOTE_VIDEO_STATS"
    let ON_AUDIO_DEVICE_CHANGED = "ON_AUDIO_DEVICE_CHANGED"

    // MARK: - Setup

    override init() {
        super.init()
    }

    override class func requiresMainQueueSetup() -> Bool {
        true
    }

    override func supportedEvents() -> [String]! {
        return [ON_JOIN, ON_PREVIEW, ON_ROOM_UPDATE, ON_PEER_UPDATE, ON_TRACK_UPDATE, ON_ERROR, ON_MESSAGE, ON_SPEAKER, RECONNECTING, RECONNECTED, ON_ROLE_CHANGE_REQUEST, ON_CHANGE_TRACK_STATE_REQUEST, ON_REMOVED_FROM_ROOM, ON_RTC_STATS, ON_LOCAL_AUDIO_STATS, ON_LOCAL_VIDEO_STATS, ON_REMOTE_AUDIO_STATS, ON_REMOTE_VIDEO_STATS, ON_AUDIO_DEVICE_CHANGED]
    }

    // MARK: - HMS SDK Actions

    @objc
    func build(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        DispatchQueue.main.async { [weak self] in
            let hasItem = self?.hmsCollection.index(forKey: "12345")

            if let _ = hasItem {
                let id = UUID().uuidString
                let hms = HMSRNSDK(data: data, delegate: self, uid: id)
                self?.hmsCollection[id] = hms

                resolve?(id)
            } else {
                let id = "12345"
                let hms = HMSRNSDK(data: data, delegate: self, uid: id)
                self?.hmsCollection[id] = hms

                resolve?(id)
            }
        }
    }

    @objc
    func preview(_ credentials: NSDictionary) {
        let hms = HMSHelper.getHms(credentials, hmsCollection)
        hms?.preview(credentials)
    }

    @objc
    func previewForRole(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.previewForRole(data, resolve, reject)
    }

    @objc
    func cancelPreview(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.cancelPreview()
    }

    @objc
    func join(_ credentials: NSDictionary) {
        let hms = HMSHelper.getHms(credentials, hmsCollection)
        hms?.join(credentials)
    }

    @objc
    func setLocalMute(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.setLocalMute(data)
    }

    @objc
    func setLocalVideoMute(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.setLocalVideoMute(data)
    }

    @objc
    func switchCamera(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.switchCamera()
    }

    @objc
    func leave(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.leave(resolve, reject)
    }

    @objc
    func sendBroadcastMessage(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.sendBroadcastMessage(data, resolve, reject)
    }

    @objc
    func sendGroupMessage(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.sendGroupMessage(data, resolve, reject)
    }

    @objc
    func sendDirectMessage(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.sendDirectMessage(data, resolve, reject)
    }

    @objc
    func acceptRoleChange(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.acceptRoleChange(resolve, reject)
    }

    @available(*, deprecated, message: "Use changeRoleOfPeer function")
    @objc
    func changeRole(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.changeRole(data, resolve, reject)
    }

    @objc
    func changeRoleOfPeer(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.changeRole(data, resolve, reject)
    }

    @objc
    func changeRoleOfPeersWithRoles(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.changeRolesOfAllPeers(data, resolve, reject)
    }

    @objc
    func changeTrackState(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.changeTrackState(data, resolve, reject)
    }

    @objc
    func changeTrackStateForRoles(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.changeTrackStateForRoles(data, resolve, reject)
    }

    @objc
    func isMute(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)
        if let hmsInstance = hms {
            hmsInstance.isMute(data, resolve, reject)
        } else {
            reject?(nil, "NO_INSTANCE", nil)
            return
        }
    }

    @objc
    func removePeer(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.removePeer(data, resolve, reject)
    }

    @objc
    func endRoom(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.endRoom(data, resolve, reject)
    }

    @objc
    func setPlaybackAllowed(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.setPlaybackAllowed(data)
    }

    @objc
    func isPlaybackAllowed(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.isPlaybackAllowed(data, resolve, reject)
    }

    @objc
    func setPlaybackForAllAudio(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.setPlaybackForAllAudio(data)
    }

    @objc
    func remoteMuteAllAudio(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.remoteMuteAllAudio(resolve, reject)
    }

    @objc
    func changeMetadata(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.changeMetadata(data, resolve, reject)
    }

    @objc
    func setVolume(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.setVolume(data)
    }

    @objc
    func startRTMPOrRecording(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.startRTMPOrRecording(data, resolve, reject)
    }

    @objc
    func stopRtmpAndRecording(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.stopRtmpAndRecording(resolve, reject)
    }

    @objc
    func startHLSStreaming(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.startHLSStreaming(data, resolve, reject)
    }

    @objc
    func stopHLSStreaming(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.stopHLSStreaming(resolve, reject)
    }

    @objc
    func changeName(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.changeName(data, resolve, reject)
    }

    @objc
    func destroy(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let id = data.value(forKey: "id") as? String ?? "12345"
        hmsCollection.removeValue(forKey: id)
        resolve?(["success": id + " removed"])
    }

    @objc
    func startScreenshare(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.startScreenshare(resolve, reject)
    }

    @objc
    func stopScreenshare(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.stopScreenshare(resolve, reject)
    }

    @objc
    func isScreenShared(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.isScreenShared(resolve, reject)
    }

    @objc
    func playAudioShare(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.playAudioShare(data, resolve, reject)
    }

    @objc
    func setAudioShareVolume(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.setAudioShareVolume(data)
    }

    @objc
    func stopAudioShare(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.stopAudioShare(data)
    }

    @objc
    func resumeAudioShare(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.resumeAudioShare(data)
    }

    @objc
    func pauseAudioShare(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.pauseAudioShare(data)
    }

    @objc
    func audioShareIsPlaying(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.audioShareIsPlaying(data, resolve, reject)
    }

    @objc
    func audioShareCurrentTime(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.audioShareCurrentTime(data, resolve, reject)
    }

    @objc
    func audioShareDuration(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.audioShareDuration(data, resolve, reject)
    }

    @objc
    func enableNetworkQualityUpdates(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.enableNetworkQualityUpdates()
    }

    @objc
    func disableNetworkQualityUpdates(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.disableNetworkQualityUpdates()
    }

    @objc
    func setSessionMetaData(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.setSessionMetaData(data, resolve, reject)
    }

    @objc
    func getPeerProperty(_ data: NSDictionary) -> [AnyHashable: Any]? {
        let hms = HMSHelper.getHms(data, hmsCollection)

        return hms?.getPeerProperty(data)
    }

    @objc
    func getRoomProperty(_ data: NSDictionary) -> [AnyHashable: Any]? {
        let hms = HMSHelper.getHms(data, hmsCollection)

        return hms?.getRoomProperty(data)
    }

    @objc
    func enableEvent(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.enableEvent(data, resolve, reject)
    }

    @objc
    func disableEvent(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.disableEvent(data, resolve, reject)
    }

    @objc
    func restrictData(_ data: NSDictionary) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.restrictData(data)
    }

    // MARK: - HMS SDK Get APIs
    @objc
    func getRoom(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.getRoom(resolve)
    }

    @objc
    func getLocalPeer(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.getLocalPeer(resolve)
    }

    @objc
    func getRemotePeers(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.getRemotePeers(resolve)
    }

    @objc
    func getRoles(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.getRoles(resolve)
    }

    @objc
    func getSessionMetaData(_ data: NSDictionary, _ resolve: RCTPromiseResolveBlock?, _ reject: RCTPromiseRejectBlock?) {
        let hms = HMSHelper.getHms(data, hmsCollection)

        hms?.getSessionMetaData(resolve, reject)
    }

    // MARK: - HMS SDK Delegate Callbacks
    func emitEvent(_ name: String, _ data: [String: Any]) {
        self.sendEvent(withName: name, body: data)
    }
}

File: packages/react-native-hms/android/src/oldarch/java/com/reactnativehmssdk/HMSManager.kt
package com.reactnativehmssdk

import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule

/**
 * HMSManager — old-arch wrapper for the HMSManager native module.
 *
 * Phase 1 / 1C-2 of the New Architecture migration. This wrapper is compiled
 * only when the consumer's app uses old architecture (newArchEnabled=false).
 * It extends ReactContextBaseJavaModule (the legacy bridge base class) and
 * exposes ~98 @ReactMethod-annotated functions that all delegate to
 * HMSManagerImpl in src/main/.
 *
 * The new-arch counterpart at src/newarch/.../HMSManager.kt extends the
 * Codegen-generated NativeHMSManagerSpec and forwards to the same impl.
 *
 * Pattern reference:
 * https://github.com/reactwg/react-native-new-architecture/blob/main/docs/backwards-compat-turbo-modules.md
 */
@ReactModule(name = HMSManagerImpl.REACT_CLASS)
class HMSManager(
  reactContext: ReactApplicationContext,
) : ReactContextBaseJavaModule(reactContext) {
  private val impl = HMSManagerImpl(reactContext)

  override fun getName(): String = HMSManagerImpl.REACT_CLASS

  /**
   * Lifecycle hooks that consumer apps' MainActivity calls. Forward to
   * the Impl's companion-object methods for backward compatibility with
   * apps that already call HMSManager.onPictureInPictureModeChanged(...)
   * etc.
   */
  companion object {
    @JvmStatic
    fun onPictureInPictureModeChanged(
      isInPictureInPictureMode: Boolean,
      newConfig: android.content.res.Configuration,
    ) = HMSManagerImpl.onPictureInPictureModeChanged(isInPictureInPictureMode, newConfig)

    @JvmStatic
    fun onResume() = HMSManagerImpl.onResume()

    @JvmStatic
    fun onUserLeaveHint() = HMSManagerImpl.onUserLeaveHint()
  }

  /** Used by view managers to look up the active SDK instances map. */
  fun getHmsInstance(): MutableMap<String, HMSRNSDK> = impl.getHmsInstance()

  @ReactMethod
  fun build(data: ReadableMap?, callback: Promise?) = impl.build(data, callback)

  @ReactMethod
  fun preview(credentials: ReadableMap) = impl.preview(credentials)

  @ReactMethod
  fun join(credentials: ReadableMap) = impl.join(credentials)

  @ReactMethod
  fun setLocalMute(data: ReadableMap) = impl.setLocalMute(data)

  @ReactMethod
  fun setLocalVideoMute(data: ReadableMap) = impl.setLocalVideoMute(data)

  @ReactMethod
  fun switchCamera(data: ReadableMap) = impl.switchCamera(data)

  @ReactMethod
  fun leave(data: ReadableMap, callback: Promise?) = impl.leave(data, callback)

  @ReactMethod
  fun sendBroadcastMessage(data: ReadableMap, callback: Promise?) = impl.sendBroadcastMessage(data, callback)

  @ReactMethod
  fun sendGroupMessage(data: ReadableMap, callback: Promise?) = impl.sendGroupMessage(data, callback)

  @ReactMethod
  fun sendDirectMessage(data: ReadableMap, callback: Promise?) = impl.sendDirectMessage(data, callback)

  @ReactMethod
  fun changeRole(data: ReadableMap, callback: Promise?) = impl.changeRole(data, callback)

  @ReactMethod
  fun changeRoleOfPeer(data: ReadableMap, promise: Promise?) = impl.changeRoleOfPeer(data, promise)

  @ReactMethod
  fun changeRoleOfPeersWithRoles(data: ReadableMap, promise: Promise?) = impl.changeRoleOfPeersWithRoles(data, promise)

  @ReactMethod
  fun changeTrackState(data: ReadableMap, callback: Promise?) = impl.changeTrackState(data, callback)

  @ReactMethod
  fun changeTrackStateForRoles(data: ReadableMap, callback: Promise?) = impl.changeTrackStateForRoles(data, callback)

  @ReactMethod
  fun isMute(data: ReadableMap, callback: Promise?) = impl.isMute(data, callback)

  @ReactMethod
  fun removePeer(data: ReadableMap, callback: Promise?) = impl.removePeer(data, callback)

  @ReactMethod
  fun isPlaybackAllowed(data: ReadableMap, callback: Promise?) = impl.isPlaybackAllowed(data, callback)

  @ReactMethod
  fun getRoom(data: ReadableMap, callback: Promise?) = impl.getRoom(data, callback)

  @ReactMethod
  fun getLocalPeer(data: ReadableMap, callback: Promise?) = impl.getLocalPeer(data, callback)

  @ReactMethod
  fun getRemotePeers(data: ReadableMap, callback: Promise?) = impl.getRemotePeers(data, callback)

  @ReactMethod
  fun getRoles(data: ReadableMap, callback: Promise?) = impl.getRoles(data, callback)

  @ReactMethod
  fun setPlaybackAllowed(data: ReadableMap, callback: Promise?) = impl.setPlaybackAllowed(data, callback)

  @ReactMethod
  fun endRoom(data: ReadableMap, callback: Promise?) = impl.endRoom(data, callback)

  @ReactMethod
  fun previewForRole(data: ReadableMap, callback: Promise?) = impl.previewForRole(data, callback)

  @ReactMethod
  fun cancelPreview(data: ReadableMap, callback: Promise?) = impl.cancelPreview(data, callback)

  @ReactMethod
  fun acceptRoleChange(data: ReadableMap, callback: Promise?) = impl.acceptRoleChange(data, callback)

  @ReactMethod
  fun setVolume(data: ReadableMap, callback: Promise?) = impl.setVolume(data, callback)

  @ReactMethod
  fun getVolume(data: ReadableMap, callback: Promise?) = impl.getVolume(data, callback)

  @ReactMethod
  fun setPlaybackForAllAudio(data: ReadableMap, callback: Promise?) = impl.setPlaybackForAllAudio(data, callback)

  @ReactMethod
  fun remoteMuteAllAudio(data: ReadableMap, callback: Promise?) = impl.remoteMuteAllAudio(data, callback)

  @ReactMethod
  fun changeMetadata(data: ReadableMap, callback: Promise?) = impl.changeMetadata(data, callback)

  @ReactMethod
  fun startScreenshare(data: ReadableMap, callback: Promise?) = impl.startScreenshare(data, callback)

  @ReactMethod
  fun isScreenShared(data: ReadableMap, callback: Promise?) = impl.isScreenShared(data, callback)

  @ReactMethod
  fun stopScreenshare(data: ReadableMap, callback: Promise?) = impl.stopScreenshare(data, callback)

  @ReactMethod
  fun startAudioshare(data: ReadableMap, callback: Promise?) = impl.startAudioshare(data, callback)

  @ReactMethod
  fun isAudioShared(data: ReadableMap, callback: Promise?) = impl.isAudioShared(data, callback)

  @ReactMethod
  fun stopAudioshare(data: ReadableMap, callback: Promise?) = impl.stopAudioshare(data, callback)

  @ReactMethod
  fun getAudioMixingMode(data: ReadableMap, callback: Promise?) = impl.getAudioMixingMode(data, callback)

  @ReactMethod
  fun setAudioMixingMode(data: ReadableMap, callback: Promise?) = impl.setAudioMixingMode(data, callback)

  @ReactMethod
  fun startRTMPOrRecording(data: ReadableMap, callback: Promise?) = impl.startRTMPOrRecording(data, callback)

  @ReactMethod
  fun stopRtmpAndRecording(data: ReadableMap, callback: Promise?) = impl.stopRtmpAndRecording(data, callback)

  @ReactMethod
  fun startHLSStreaming(data: ReadableMap, callback: Promise?) = impl.startHLSStreaming(data, callback)

  @ReactMethod
  fun stopHLSStreaming(data: ReadableMap, callback: Promise?) = impl.stopHLSStreaming(data, callback)

  @ReactMethod
  fun sendHLSTimedMetadata(data: ReadableMap, callback: Promise?) = impl.sendHLSTimedMetadata(data, callback)

  @ReactMethod
  fun changeName(data: ReadableMap, callback: Promise?) = impl.changeName(data, callback)

  @ReactMethod
  fun destroy(data: ReadableMap, callback: Promise?) = impl.destroy(data, callback)

  @ReactMethod
  fun enableNetworkQualityUpdates(data: ReadableMap) = impl.enableNetworkQualityUpdates(data)

  @ReactMethod
  fun disableNetworkQualityUpdates(data: ReadableMap) = impl.disableNetworkQualityUpdates(data)

  @ReactMethod
  fun getAudioDevicesList(data: ReadableMap, callback: Promise?) = impl.getAudioDevicesList(data, callback)

  @ReactMethod
  fun getAudioOutputRouteType(data: ReadableMap, callback: Promise?) = impl.getAudioOutputRouteType(data, callback)

  @ReactMethod
  fun switchAudioOutput(data: ReadableMap, callback: Promise?) = impl.switchAudioOutput(data, callback)

  @ReactMethod
  fun setAudioMode(data: ReadableMap, callback: Promise?) = impl.setAudioMode(data, callback)

  @ReactMethod
  fun setAudioDeviceChangeListener(data: ReadableMap) = impl.setAudioDeviceChangeListener(data)

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun getPeerProperty(data: ReadableMap): WritableMap? = impl.getPeerProperty(data)

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun getRoomProperty(data: ReadableMap): WritableMap? = impl.getRoomProperty(data)

  @ReactMethod
  fun enableEvent(data: ReadableMap, promise: Promise?) = impl.enableEvent(data, promise)

  @ReactMethod
  fun disableEvent(data: ReadableMap, promise: Promise?) = impl.disableEvent(data, promise)

  @ReactMethod
  fun restrictData(data: ReadableMap, promise: Promise?) = impl.restrictData(data, promise)

  @ReactMethod
  fun getAuthTokenByRoomCode(data: ReadableMap, promise: Promise) = impl.getAuthTokenByRoomCode(data, promise)

  @ReactMethod
  fun handlePipActions(action: String, data: ReadableMap, promise: Promise?) = impl.handlePipActions(action, data, promise)

  @ReactMethod
  fun getRemoteVideoTrackFromTrackId(data: ReadableMap, promise: Promise) = impl.getRemoteVideoTrackFromTrackId(data, promise)

  @ReactMethod
  fun getRemoteAudioTrackFromTrackId(data: ReadableMap, promise: Promise) = impl.getRemoteAudioTrackFromTrackId(data, promise)

  @ReactMethod
  fun getVideoTrackLayer(data: ReadableMap, promise: Promise) = impl.getVideoTrackLayer(data, promise)

  @ReactMethod
  fun getVideoTrackLayerDefinition(data: ReadableMap, promise: Promise) = impl.getVideoTrackLayerDefinition(data, promise)

  @ReactMethod
  fun setVideoTrackLayer(data: ReadableMap, promise: Promise?) = impl.setVideoTrackLayer(data, promise)

  @ReactMethod
  fun captureImageAtMaxSupportedResolution(data: ReadableMap, promise: Promise?) = impl.captureImageAtMaxSupportedResolution(data, promise)

  @ReactMethod
  fun setSessionMetadataForKey(data: ReadableMap, promise: Promise?) = impl.setSessionMetadataForKey(data, promise)

  @ReactMethod
  fun getSessionMetadataForKey(data: ReadableMap, promise: Promise?) = impl.getSessionMetadataForKey(data, promise)

  @ReactMethod
  fun addKeyChangeListener(data: ReadableMap, promise: Promise?) = impl.addKeyChangeListener(data, promise)

  @ReactMethod
  fun removeKeyChangeListener(data: ReadableMap, promise: Promise?) = impl.removeKeyChangeListener(data, promise)

  @ReactMethod
  fun getRoomLayout(data: ReadableMap, promise: Promise?) = impl.getRoomLayout(data, promise)

  @ReactMethod
  fun raiseLocalPeerHand(data: ReadableMap, promise: Promise?) = impl.raiseLocalPeerHand(data, promise)

  @ReactMethod
  fun lowerLocalPeerHand(data: ReadableMap, promise: Promise?) = impl.lowerLocalPeerHand(data, promise)

  @ReactMethod
  fun lowerRemotePeerHand(data: ReadableMap, promise: Promise?) = impl.lowerRemotePeerHand(data, promise)

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun getPeerListIterator(data: ReadableMap): WritableMap? = impl.getPeerListIterator(data)

  @ReactMethod
  fun peerListIteratorHasNext(data: ReadableMap, promise: Promise?) = impl.peerListIteratorHasNext(data, promise)

  @ReactMethod
  fun peerListIteratorNext(data: ReadableMap, promise: Promise?) = impl.peerListIteratorNext(data, promise)

  @ReactMethod
  fun checkNotifications(promise: Promise?) = impl.checkNotifications(promise)

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun setSoftInputMode(inputMode: Int): Int = impl.setSoftInputMode(inputMode)

  @ReactMethod(isBlockingSynchronousMethod = true)
  fun getSoftInputMode(): Int = impl.getSoftInputMode()

  @ReactMethod
  fun hideSystemBars() = impl.hideSystemBars()

  @ReactMethod
  fun showSystemBars() = impl.showSystemBars()

  @ReactMethod
  fun setAlwaysScreenOn(data: ReadableMap, promise: Promise?) = impl.setAlwaysScreenOn(data, promise)

  @ReactMethod
  fun quickStartPoll(data: ReadableMap, promise: Promise?) = impl.quickStartPoll(data, promise)

  @ReactMethod
  fun addResponseOnPollQuestion(data: ReadableMap, promise: Promise?) = impl.addResponseOnPollQuestion(data, promise)

  @ReactMethod
  fun stopPoll(data: ReadableMap, promise: Promise?) = impl.stopPoll(data, promise)

  @ReactMethod
  fun fetchLeaderboard(data: ReadableMap, promise: Promise?) = impl.fetchLeaderboard(data, promise)

  @ReactMethod
  fun startWhiteboard(data: ReadableMap, promise: Promise?) = impl.startWhiteboard(data, promise)

  @ReactMethod
  fun stopWhiteboard(data: ReadableMap, promise: Promise?) = impl.stopWhiteboard(data, promise)

  @ReactMethod
  fun enableNoiseCancellationPlugin(data: ReadableMap, promise: Promise?) = impl.enableNoiseCancellationPlugin(data, promise)

  @ReactMethod
  fun disableNoiseCancellationPlugin(data: ReadableMap, promise: Promise?) = impl.disableNoiseCancellationPlugin(data, promise)

  @ReactMethod
  fun isNoiseCancellationPluginEnabled(data: ReadableMap, promise: Promise?) = impl.isNoiseCancellationPluginEnabled(data, promise)

  @ReactMethod
  fun isNoiseCancellationPluginAvailable(data: ReadableMap, promise: Promise?) = impl.isNoiseCancellationPluginAvailable(data, promise)

  @ReactMethod
  fun handleRealTimeTranscription(data: ReadableMap, promise: Promise?) = impl.handleRealTimeTranscription(data, promise)

  @ReactMethod
  fun setPermissionsAccepted(data: ReadableMap, promise: Promise?) = impl.setPermissionsAccepted(data, promise)

  @ReactMethod
  fun addListener(eventName: String) = impl.addListener(eventName)

  @ReactMethod
  fun removeListeners(count: Int) = impl.removeListeners(count)

}
